### DISCUSSION TITLE
Run DCR in Fastly Edge Compute

### DISCUSSION THREAD

**Discussion by oliverlloyd** [guardian/source#1233] [Ideas]
# DCR in Fastly Edge Compute

## What?
Run DCR - a node server - as a function in Fastly Edge Compute.

https://docs.fastly.com/products/compute-at-edge

Instead of having a farm of EC2 servers that provide a rendering service for requests that get routed to origin, we move that logic closer to the browser.

To achieve this the request would still need to get CAPI data, this would mean that instead of making the request for the article CAPI data inside Frontend, it would be Fastly making the request and it is _this_ request that Fastly would cache.

### Performance
The request for CAPI data is much more cachable than the one for the article html. We vary the html response based on a number of things right now (AB tests, region, etc) but the CAPI data does not vary at all (is this true?). This means far fewer origin requests so faster response times.

The flip side to this is that every request will be 'rendered' by DCR so the time taken to render *must* be fast. The sell on Fastly's edge cache and the reports from other people are that this is the case but this would need careful testing and monitoring.

### Flexibility
Right now we carefully, _reluctantly_ split the cache so provide a varied response to different types of request. We want to do this more, we want a more customised response to readers based on their individual state, but providing this is expensive, complex, and does not scale.

But if we put DCR at the edge then we can provide a much higher level of customisation with zero added cost. In addition the code required to do this would be a lot simpler and easier to reason about and, crucially, much easier to test. Instead of complex conditional logic living in different platforms, the conditionals needed to decide what a user is served will all live in one place, DCR.

## But DCR isn't a web server
Right now DCR only accepts a POST request from Frontend. It lives behind Frontend as a rendering service.

To make this idea work you would need to make DCR a web server that accepts a get request. This is simple to do on the DCR side of things but would need some routing work on Frontend

**Comment by sndrs** [comment]
we'd need a careful look at what frontend does – iirc it does a lot more that routing e.g. header management etc.

at this point moving to something like next.js maybe makes more sense than effectively writing the guardian's own version of next. i'm personally tempted by that proposition, but i wonder if we can get edge to intervene without actually needing to render every page in full?

**Comment by sndrs** [comment]
also, did you mean to raise this in source repo?

**Reply by oliverlloyd** [reply]
We don't have Discussions enabled in DCR and this seemed like the place where discussions are happening... Is there a more appropriate location? I find not knowing where stuff goes is a barrier to adopting some Github features

**Comment by arelra** [comment]
Interesting idea!

aiui if DCR renders at-edge this would happen post cache and hence the time taken to render would be added to all requests?

Some points/questions:

1. Rendering at edge could increase TTFB and potentially introduce more _variability_ into response times - perhaps by a very small amount but something to be wary of.
1. Is there a hybrid model where the stateless render can return from the cache but stateful/personalised renders can ~SSR~ er, ESR (Edge Side Render).
1. This maybe heresy but if we can ESR every request and it's within cost, devex & performance bounds - why cache at all?

**Reply by sndrs** [reply]
yeah i'm intrigued by 3 too. especially if it can timeout (so ideally e.g. user data is already in the page, but can be requested later if the request takes too long)

**Reply by oliverlloyd** [reply]
On 1 (Rendering at edge could increase TTFB). It does add time for sure and this is potentially a problem. The real worry for me is right now we can write code that makes the server render slower but we're not _too_ concerned by this. We have the mindset that the result is cached.

But this changes DCR to being a fully user facing service which changes the dynamic of how we should think about out code. Arguably we should be thinking about it this way anyway...

Worth noting that Fastly compiles the code to WASM and advertises near zero startup times so it could be we're talking about very small times here

**Reply by groakland** [reply]
[Lucet](https://www.fastly.com/blog/announcing-lucet-fastly-native-webassembly-compiler-runtime) is Fastly's  open source WASM compiler and runtime, boasting startup times <50µs : 

_"A major design requirement for Lucet was to be able to execute on every single request that Fastly handles. That means creating a WebAssembly instance for each of the tens of thousands of requests per second in a single process, which requires a dramatically lower runtime footprint than possible with a browser JavaScript engine. Lucet can instantiate WebAssembly modules in under 50 microseconds, with just a few kilobytes of memory overhead. By comparison, Chromium’s V8 engine takes about 5 milliseconds, and tens of megabytes of memory overhead, to instantiate JavaScript or WebAssembly programs."_

**Reply by arelra** [reply]
> We have the mindset that the result is cached.

> But this changes DCR to being a fully user facing service which changes the dynamic of how we should think about out code. Arguably we should be thinking about it this way anyway...

This seems like a good dev pressure to have!

Even for the current architecture, when Fastly goes to origin there is no hiding behind the cache.

Does DCR log a sampling of server render times (i.e pure SSR times with no network trips) to the data lake or anywhere else?

That would be really interesting data to have, and to have it broken down by various metrics e.g. article type, length etc.

> Worth noting that Fastly compiles the code to WASM and advertises near zero startup times so it could be we're talking about very small times here

They sell start up time quite hard and the demos are impressive.

It will be interesting to compare compute performance vs dedicated EC2 instances.

**Reply by oliverlloyd** [reply]
> Does DCR log a sampling of server render times

Not yet but we've got [a PR to start logging dev build](https://github.com/guardian/dotcom-rendering/pull/3848) times and once that's up and running we can look at extending that solution to other metrics

**Comment by oliverlloyd** [comment]
## Why cache at all?

Creating a new thread for this idea from @arelra 

> This maybe heresy but if we can ESR every request and it's within cost, devex & performance bounds - why cache at all?

**Reply by oliverlloyd** [reply]
To my mind this makes sense. It's easier to change something completely rather than add complexity.

We would still need to cache the CAPI call to origin. There are worlds where distributed storage is possible but I'm not sure we want to go there just yet (or maybe we do?)

**Reply by arelra** [reply]
The cache is obviously a fantastic tool that works well for us, but it's also something of a straight-jacket too.

There's always a trade off but imo the less tie in to custom Fastly solutions at edge the better e.g. dbs, KV stores etc.

If DCR can construct all its external needs as API calls to origin(s) (that get cached) that would allow us to maintain devex and also the slight benefit of less buy in to any particular edge provider.

**Reply by oliverlloyd** [reply]
I think this discussion deserves a reboot in light of potential new requirements for personalisation.

I am very much ambivalent here and these statements really are meant to springboard other ideas rather than be seen as targets to knock over 🙏 . I also think any decision will be a trade off but it seems clear that we have new motivations that could, potentially, make these sorts of ideas more viable so maybe it's good to captcha some thoughts

## Considerations

### Cost
This is math but it could be there are solutions that meet our requirements that save money? Or perhaps cost more but gives us greater wins in other areas?

### Resilience
We think of the cache as a safety net to save us when there are 

### Images
Would we continue to use Fastly IO?

### Dev ex
Should we have some code running at origin and some at the edge? Would a fully edge solution be simpler?
